### PR TITLE
Fix deadlock on multiple peers communicating 

### PR DESCRIPTION
**Pull request by MarcelTorne**
For the moment I have fixed the deadlock but we should verify wether it is always waiting for all of the weights to arrive or if it already detects if all weights receive before the 10 seconds.

### CODE REVIEW

**Comment by PaulMansat** [comment]
Thanks for noticing this issue Marcel! 
Maybe it's a bit strange to wait for data to arrive when checking the array's length. Maybe we can return the array's length all together without waiting for data to arrive (we are doing the waiting with the function dataReceivedBreak). 
I'll have a look into it tomorrow. 
Thanks again, I hope you are now able to train with 4 peers 👍.

**Comment by MarcelTorne** [comment]
> Thanks for noticing this issue Marcel!
> Maybe it's a bit strange to wait for data to arrive when checking the array's length. Maybe we can return the array's length all together without waiting for data to arrive (we are doing the waiting with the function dataReceivedBreak).
> I'll have a look into it tomorrow.
> Thanks again, I hope you are now able to train with 4 peers 👍.

Yeah that's true, my thing is just a quick fix. But the problem with just returning the arrays length is that once you receive one weight it will directly continue to averaging. So there should still be a break until you receive `threshold` weights or 10 seconds elapsed. What do you think?

**Comment by MarcelTorne** [comment]
I found the other problem we had, on `onEpochEndCommon` we use create a copy of the average weights, so once it was passed into the function it would never change again since it was a copy of the array not a pointer to the recvBuffer object. Please let me know what do you think of the fix and wether you have some better ideas.

File: mobile-browser-based-version/src/helpers/communication_script/helpers.js [hunk 1]
```diff
@@ -116,18 +117,38 @@ export function dataReceivedBreak(recvBuffer, key, maxTries) {
...
+    });
+}
+
+/**
+ * Waits until an array reaches a given length. Used to make 
+ * sure that all weights from peers are received.
+ * @param {Array} arr 
+ * @param {int} len 
+ */
+ export function checkArrayLenCommon(recvBuffer, len) {
```
Comment (mmilenkoski) [inline mobile-browser-based-version/src/helpers/communication_script/helpers.js:138]: This seems like code duplication to me. Is it possible to have only `checkArrayLenSync` and call it with `Object.values(recvBuffer.avgWeights).flat(1)` to achieve the functionality of `checkArrayLenCommon` or have a flag that indicates if you need to call `Object.values(recvBuffer.avgWeights).flat(1)` inside?

File: mobile-browser-based-version/src/task_definition/lus_covid.js [hunk 2]
```diff
@@ -280,7 +280,7 @@ export const trainingInformation = {
     modelTrainData: {
         epochs: 10,
     },
-    threshold: 1,
+    threshold: 2,
```
Comment (mmilenkoski) [inline mobile-browser-based-version/src/task_definition/lus_covid.js:283]: is this change intentional?

File: mobile-browser-based-version/src/helpers/communication_script/helpers.js [hunk 3]
```diff
@@ -116,18 +117,38 @@ export function dataReceivedBreak(recvBuffer, key, maxTries) {
  * @param {Array} arr 
  * @param {int} len 
  */
-export function checkArrayLen(arr, len) {
+export function checkArrayLenSync(arr, len) {
+    let maxTries = 100
```
Comment (mmilenkoski) [inline mobile-browser-based-version/src/helpers/communication_script/helpers.js:121]: please make this a constant on top of the file

**Review by mmilenkoski — CHANGES_REQUESTED** [review]

File: mobile-browser-based-version/src/task_definition/lus_covid.js (hunk 2 above)
Comment (MarcelTorne) [reply mobile-browser-based-version/src/task_definition/lus_covid.js:283]: Well, it depends on how many peers we want to wait for, so I increased in for the benchmarks. However, I suppose that in the future we can either adapt it dynamically or let the user set it in the interface. I'll still set it back to 1.

File: mobile-browser-based-version/src/helpers/communication_script/helpers.js (hunk 1 above)
Comment (MarcelTorne) [reply mobile-browser-based-version/src/helpers/communication_script/helpers.js:139]: Before we used to have only `checkArrayLenSync` and we called it with `Object.values(recvBuffer.avgWeights).flat(1)` but the problem is that this is not a pointer to the `recvBuffer.avgWeights` anymore but it's a copy so its value will never change and hence even if a new weight is received you will not detect and you will wait for 10 seconds.

Comment (MarcelTorne) [reply mobile-browser-based-version/src/helpers/communication_script/helpers.js:139]: I can add the flag but I don't really like it with the flag, and I don't have any better idea for a relatively quick fix. Do you?

Comment (mmilenkoski) [reply mobile-browser-based-version/src/helpers/communication_script/helpers.js]: okay, I understand. Why don't you like it with a flag though? In my opinion, it is better than having two functions with 95% duplicated code.

Comment (MarcelTorne) [reply mobile-browser-based-version/src/helpers/communication_script/helpers.js]: Yes definitely, I already changed it before, but I was asking wether you would have a better idea, sorry for the misunderstanding.

**Comment by PaulMansat** [comment]
Hey @MarcelTorne, could you resolve the conflict on this branch? I think it's simply about making the choice of keeping or net the check array function in sync. Thanks! 
When this is done, we can merge the branch to close all this PR because there are 2 other big PR to merge (one changing how files are uploaded and the other changing the routing architecture). Thanks!

**Review by PaulMansat — APPROVED** [review]
